File: c/src/rational.c
// Author: Ratnodeep Bandyopadhyay Copyright. 2021
#include "rational.h"

int mod_div(struct rational* x) {
    
    int quo = x->num / x->den;
    int mod = x->num % x->den;

    if(mod >= (x->den / 2)) {
        return quo + 1;
    } else {
        return quo;
    }
}

void rat_add_rat(struct rational* l, struct rational* r, struct rational* s) {
    
    if(l->den == r->den) {
        s->num = l->num + r->num;
        s->den = l->den;
    } else {
        s->num = (l->num * r->den) + (r->num * l->den);
        s->den = l->den * r->den;
    }

    return;
}

void rat_mul_rat(struct rational* l, struct rational* r, struct rational* s) {
    s->num = l->num * r->num;
    s->den = l->den * r->den;
    return;
}

void rat_div_rat(struct rational* l, struct rational* r, struct rational* s) {
    s->num = l->num * r->den;
    s->den = l->den * r->num;
    return;
}

File: python/babai/src/rational.py
# Author: Ratnodeep Bandyopadhyay. Copyright 2021. All rights reserved.

class Rational:

    def __init__(self, num, den):
        self.num = num
        self.den = den

    def __repr__(self):
        return "{} / {}".format(self.num, self.den)

    @staticmethod
    def mod_round(self):

        if(self.den < 0):
            self.num = -self.num
            self.den = -self.den
            
        quo = self.num/self.den
        rem = self.num % self.den

        if(rem >= self.den / 2):
            return quo + 1
        else:
            return quo

    @staticmethod
    def rat_add_rat(l, r, s):
        if(l.den != r.den):
            s.num = (l.num * r.den) + (r.num * l.den)
            s.den = l.den * r.den
        else:
            s.num = l.num + r.num
            s.den = l.den
        return

    @staticmethod
    def rat_mul_rat(l, r, s):
        s.num = l.num * r.num
        s.den = l.den * r.den
        return


    @staticmethod
    def rat_div_rat(l, r, s):
        s.num = l.num * r.den
        s.den = l.den * r.num
        return
